File: ruby/700-lambda.rb
ZERO = ->(f) { ->(v) {       v   } }
ONE  = ->(f) { ->(v) {    f.(v)  } }
TWO  = ->(f) { ->(v) { f.(f.(v)) } }

DECREMENT = ->(n) { ->(f) { ->(x) {
  n.(->(g) { ->(h) { h.(g.(f)) } }).(->(y) { x }).(->(y){ y })
} } }

MULT = ->(a) { ->(b) { ->(f) {  a.(b.(f)) } } }
EIGHT = MULT.(TWO).(MULT.(TWO).(TWO))

TRU = ->(x) { ->(y) { x } }
FLSE = ->(x) { ->(y) { y } }

IS_ZERO = ->(f) { f.(->(v) { FLSE }).(TRU)  }

IF = ->(bool) { ->(a) { ->(b) { bool.(a).(b)  } } }

Z = ->(z) {
  ->(f) {    ->(n){ f.(f).(n) }  }.(
  ->(f) { z.(->(n){ f.(f).(n) }  )})
}

fact = Z.(
  ->(recursion) {
    ->(x) {
      if x == 1
        1
      else
        x * recursion.(x-1)
      end
    }
  }
)

puts fact.(8)

def to_integer(num)
  num.(->(x) { x + 1}).(0)
end


# Just substitute all the stuff in right?

#puts to_integer(FACT.(EIGHT))

File: js/700-lambda.js
const ZERO = (f) => { return (v) => { return     v   }}
const ONE  = (f) => { return (v) => { return   f(v)  }}
const TWO  = (f) => { return (v) => { return f(f(v)) }}

const DECREMENT = (n) => {
  return (f) => { return (x) => {
    return n((g) => { return (h) => { return h(g(f))}})((y) => { return x})((y) => { return y })
  }}}

const MULT = (a) => { return (b) => { return (f) => { return a(b(f)) } } }

const FOUR = MULT(TWO)(TWO)
const EIGHT = MULT(FOUR)(TWO)

const TRUE  = (x) => { return (y) => { return x } }
const FALSE = (x) => { return (y) => { return y } }

const IS_ZERO = (f) => { return f((v) => { return FALSE })(TRUE) }

const IF = (bool) => {  return bool }

const Z = (z) =>{
  return (f) => { return f(f)}
        ((f) => { return z((n) => { return f(f)(n)}) })
}

const FACT = Z(
  (recur) => {
    return (x) => {
      if (x === 0) {
        return 1
      } else {
        return x * recur(x-1)
      }
    }
  }
)

function toInt(x) {
  return x((v) => { return v + 1 })(0)
}
// Just substitute all the stuff in right?

//console.log(toInt(FACT(EIGHT)))
